### PR TITLE
Fix /var/log and /tmp mount volume mounts to show up in container

### PR DESCRIPTION
**Pull request by rhatdan**
Created new function move_mounts which will handle moving all bind mounts off
of rootfs onto temporary file system.
Mounts a tmpfs on the path, and then moves the mount points back.
We want to use this for /run, /var/log and /tmp file systems.

### CODE REVIEW

File: src/systemdhook.c [hunk 1]
```diff
@@ -332,57 +332,73 @@ static int move_mount_to_runtmp(const char *rootfs, const char *run_tmp_dir, con
...
-	}
 
-	/* Move other user specified mounts under /run to temporary directory */
-	for (unsigned i = 0; i < len; i++) {
-		/* Match destinations that begin with /run/ */
-		if (!strncmp(RUN_PREFIX, config_mounts[i], strlen(RUN_PREFIX))) {
-			if (move_mount_to_runtmp(rootfs, run_tmp_dir, config_mounts[i]) < 0) {
-				pr_perror("Failed to move %s to %s", config_mounts[i], run_tmp_dir);
+		/* Special case for /run/secrets which will not be in the 
+		   config_mounts */
```
Comment (mrunalp) [inline src/systemdhook.c]: We need to retain the check to see if /run/secrets is a mount point before we attempt to move it. It was there in the eariler code.

Comment (rhatdan) [inline src/systemdhook.c]: This was breaking. I could not see why.  The stat command was always return the st_device as 0.

You are worried about the case where /run/secrets is on /run?

File: src/systemdhook.c [hunk 2]
```diff
@@ -583,6 +523,30 @@ static int prestart(const char *rootfs,
...
+		} else {
+			rc = asprintf(&options, "mode=1777%s,context=\"%s\"", memory_str, mount_label);
+		}
+		if (rc < 0) {
+			pr_perror("Failed to allocate memory for context");
+			return -1;
+		}
+
+		/* Mount tmpfs at /tmp for systemd */
+		move_mounts(rootfs, "/tmp", config_mounts, config_mounts_len, options);
```
Comment (mrunalp) [inline src/systemdhook.c]: We do need return code checks here as Giuseppe pointed out.

**Comment by mrunalp** [comment]
Latest version of hook is failing when testing with docker.

**Comment by giuseppe** [comment]
Tested 8c9b7ccab90948c67cb4eb2f689fa8e3995b0ec0.  It works on runc

**Comment by mrunalp** [comment]
LGTM

**Comment by hogarthj** [comment]
Note this has broken my production RHEL7 docker builds

Details are here:
https://bugzilla.redhat.com/show_bug.cgi?id=1419040

The breakage is reproducible on Fedora with the oci-systemd-hook in koji too.

Incidentally I'm very surprised  that a build was pushed to stable RHEL Extras when that git commit hadn't even arrived in Fedora yet - particularly one with major change to behaviour.

**Comment by robert-scheck** [comment]
Aside of what @hogarthj says, I am also seeing since this update SELinux AVC denied messages, see https://bugzilla.redhat.com/show_bug.cgi?id=1423502 for details.

**Comment by TomSweeneyRedHat** [comment]
Fixed in oci-systemd-hook-0.1.6-1.gitfe22236

Test Results:

[root@rhel73bz ~]# cat > Dockerfile.mariadb << EOF
> FROM centos:latest
> STOPSIGNAL SIGRTMIN+3
>  
> RUN yum -y install mariadb-server && yum clean all
>  
> RUN systemctl enable mariadb
>  
> VOLUME /var/lib/mysql
>  
> CMD ["/sbin/init"]
> EOF
[root@rhel73bz ~]# 
[root@rhel73bz ~]# docker volume create --name localtest-mdb
localtest-mdb
[root@rhel73bz ~]# docker build -f Dockerfile.mariadb -t localtest-mdb .
Sending build context to Docker daemon 14.85 kB
Step 1 : FROM centos:latest
 ---> 98d35105a391
Step 2 : STOPSIGNAL SIGRTMIN+3
 ---> Using cache
 ---> d21037da37ed
Step 3 : RUN yum -y install mariadb-server && yum clean all
 ---> Using cache
 ---> 4440f237aa2b
Step 4 : RUN systemctl enable mariadb
 ---> Using cache
 ---> 007cbeb7dcff
Step 5 : VOLUME /var/lib/mysql
 ---> Using cache
 ---> dc7109037dbf
Step 6 : CMD /sbin/init
 ---> Using cache
 ---> 53e981575dd6
Successfully built 53e981575dd6
[root@rhel73bz ~]# docker run -dt -v localtest-mdb:/var/lib/mysql --name localtest-mdb localtest-mdb
8f5f9735e44d3b517c009d7ee7e6a234ea701da21b70b1308756df38cdbfb53e
[root@rhel73bz ~]# docker exec -t localtest-mdb /bin/bash -c 'for i in {1..30}; do if systemctl is-active mariadb ; then break  ; else sleep 1 ; fi done;'
inactive
inactive
activating
activating
activating
activating
active
[root@rhel73bz ~]# docker exec -t localtest-mdb mysql -e "GRANT ALL PRIVILEGES ON *.* TO 'testuser'@'%' IDENTIFIED BY 'testpassword' WITH GRANT OPTION;"
[root@rhel73bz ~]# docker stop localtest-mdb
localtest-mdb
[root@rhel73bz ~]# docker rm localtest-mdb
localtest-mdb
[root@rhel73bz ~]# uname -a
Linux rhel73bz.localdomain 3.10.0-625.el7.x86_64 #1 SMP Thu Mar 23 11:04:30 EDT 2017 x86_64 x86_64 x86_64 GNU/Linux
[root@rhel73bz ~]# cat /proc/version
Linux version 3.10.0-625.el7.x86_64 (mockbuild@x86-037.build.eng.bos.redhat.com) (gcc version 4.8.5 20150623 (Red Hat 4.8.5-11) (GCC) ) #1 SMP Thu Mar 23 11:04:30 EDT 2017
[root@rhel73bz ~]# rpm -qa | grep oci-systemd-hook
oci-systemd-hook-0.1.6-1.gitfe22236.el7.x86_64

File: src/systemdhook.c [hunk 3]
```diff
@@ -332,56 +330,75 @@ static int move_mount_to_runtmp(const char *rootfs, const char *run_tmp_dir, con
...
 
-	/* Move /run/secrets to temporary directory if it exists and it is a mountpoint */
-	if (stat(run_dir, &run_st) == 0 &&
-	    stat(run_secrets_dir, &secrets_st) == 0 &&
-	    secrets_st.st_rdev != run_st.st_rdev &&
-	    (S_ISDIR(secrets_st.st_mode))) {
-		if (move_mount_to_runtmp(rootfs, run_tmp_dir, "/run/secrets") < 0) {
-			pr_perror("Failed to move %s to %s", run_secrets_dir, run_tmp_dir);
+	/* Create the PATH directory */
+	if (!contains_mount(config_mounts, config_mounts_len, path)) {
```
Comment (wking) [inline src/systemdhook.c]: Why is this only guarding part of the `move_mounts` logic?  It looks like the only logic outside the guard is creating and removing `tmp_dir`.  But why bother with that by itself?  Can we restructure `move_mounts` to lead with:

```c
if (contains_mount(config_mounts, config_mounts_len, path)) {
  return 0;
}
```

Then after that we can have all the `tmp_dir` creation, moving, and `tmp_dir` removal without needing a further guard.

Comment (rhatdan) [reply src/systemdhook.c]: Excellent point.

**Pull request merged**